src/gradient.jl

```diff
@@ -0,0 +1,199 @@
+#
+# `multi_calc_gradient!` is just like `_calc_gradient!`, but uses Julia's multithreading
+# capabilities when performing calculations in chunk-mode.
+#
+# `VEC_MODE_EXPR` is a constant expression for vector-mode that provides the function body
+# for `calc_gradient!` and `multi_calc_gradient!` when chunk size equals input length.
+#
+# `calc_gradient_expr` takes in a vector-mode expression body or chunk-mode expression body
+# and returns a completed function body with the input body injected in the correct place.
+
+@generated function call_gradient!{S,N,L,M}(f, output::Vector{S}, x::Vector, ::Type{Val{N}}, ::Type{Val{L}}, ::Type{Val{M}})
+    gradf! = M ? :multi_calc_gradient! : :calc_gradient!
+    return :($(gradf!)(f, output, x, Val{$(N == nothing ? pick_chunk(L) : N)}, Val{L})::Tuple{S, Vector{S}})
+end
+
+@generated function calc_gradient!{S,T,N,L}(f, output::Vector{S}, x::Vector{T}, ::Type{Val{N}}, ::Type{Val{L}})
+    if N == L
+        body = VEC_MODE_EXPR
+    else
+        remainder = L % N == 0 ? N : L % N
+        fill_length = L - remainder
+        reseed_partials = remainder == N ? :() : :(seed_partials = cachefetch!(tid, Partials{N,T}, Val{$(remainder)}))
+        body = quote
+            workvec::Vector{DiffNumber{N,T}} = cachefetch!(tid, DiffNumber{N,T}, Val{L})
+            pzeros = zero(Partials{N,T})
+
+            @simd for i in 1:L
+                @inbounds workvec[i] = DiffNumber{N,T}(x[i], pzeros)
+            end
+
+            for c in 1:$(N):$(fill_length)
+                @simd for i in 1:N
+                    j = i + c - 1
+                    @inbounds workvec[j] = DiffNumber{N,T}(x[j], seed_partials[i])
+                end
+                local result::DiffNumber{N,S} = f(workvec)
+                @simd for i in 1:N
+                    j = i + c - 1
+                    @inbounds output[j] = partials(result, i)
+                    @inbounds workvec[j] = DiffNumber{N,T}(x[j], pzeros)
+                end
+            end
+
+            # Performing the final chunk manually seems to triggers some additional
+            # optimization heuristics, which results in more efficient memory allocation
+            $(reseed_partials)
+            @simd for i in 1:$(remainder)
+                j = $(fill_length) + i
+                @inbounds workvec[j] = DiffNumber{N,T}(x[j], seed_partials[i])
+            end
+            result::DiffNumber{N,S} = f(workvec)
+            @simd for i in 1:$(remainder)
+                j = $(fill_length) + i
+                @inbounds output[j] = partials(result, i)
+                @inbounds workvec[j] = DiffNumber{N,T}(x[j], pzeros)
+            end
+        end
+    end
+    return calc_gradient_expr(body)
+end
+
+if VERSION >= THREAD_VERSION
+    @generated function multi_calc_gradient!{S,T,N,L}(f, output::Vector{S}, x::Vector{T}, ::Type{Val{N}}, ::Type{Val{L}})
+        if N == L
+            body = VEC_MODE_EXPR
+        else
+            nthreads = Threads.nthreads()
```

Comment:
These need to be either `Base.Threads` or a `using Base.Threads` is needed, I think.